protocol/app/vote_extensions/extend_vote_handler.go

```diff
@@ -0,0 +1,59 @@
+package vote_extensions
+
+import (
+	"fmt"
+
+	cometabci "github.com/cometbft/cometbft/abci/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/dydxprotocol/v4-chain/protocol/app/process"
+	prices "github.com/dydxprotocol/v4-chain/protocol/x/prices/types"
+)
+
+// ExtendVoteHandler is a wrapper around the Slinky ExtendVoteHandler. This wrapper is responsible for
+// applying the newest MarketPriceUpdates in a block so that the prices to be submitted in a vote extension are
+// determined on the latest available information.
+type ExtendVoteHandler struct {
+	SlinkyExtendVoteHandler sdk.ExtendVoteHandler
+	PricesTxDecoder         process.UpdateMarketPriceTxDecoder
+	PricesKeeper            PricesKeeper
+}
+
+// ExtendVoteHandler returns a sdk.ExtendVoteHandler, responsible for the following:
+//  1. Decoding the x/prices MsgUpdateMarketPrices in the current block - fail on errors
+//  2. Validating the proposed MsgUpdateMarketPrices in accordance with the ProcessProposal check
+//  3. Updating the market prices in the PricesKeeper so that the GetValidMarketPriceUpdates function returns the
+//     latest available market prices
+//  4. Calling the Slinky ExtendVoteHandler to handle the rest of ExtendVote
+func (e *ExtendVoteHandler) ExtendVoteHandler() sdk.ExtendVoteHandler {
+	return func(ctx sdk.Context, req *cometabci.RequestExtendVote) (resp *cometabci.ResponseExtendVote, err error) {
+		// Decode the x/prices txn in the current block
+		updatePricesTx, err := e.PricesTxDecoder.DecodeUpdateMarketPricesTx(ctx, req.Txs)
+		if err != nil {
+			return nil, fmt.Errorf("DecodeMarketPricesTx failure %w", err)
+		}
+
+		// ensure that the proposed MsgUpdateMarketPrices is valid in accordance w/ stateful information
+		// this check is equivalent to the check in ProcessProposal (indexPriceCache has not been updated)
+		err = updatePricesTx.Validate()
+		if err != nil {
+			return nil, fmt.Errorf("updatePricesTx.Validate failure %w", err)
+		}
+
+		// Update the market prices in the PricesKeeper, so that the GetValidMarketPriceUpdates
+		// function returns the latest available market prices.
+		updateMarketPricesMsg, ok := updatePricesTx.GetMsg().(*prices.MsgUpdateMarketPrices)
+		if !ok {
+			return nil, fmt.Errorf("expected %s, got %T", "MsgUpdateMarketPrices", updateMarketPricesMsg)
+		}
+
+		// Update the market prices in the PricesKeeper
+		err = e.PricesKeeper.UpdateMarketPrices(ctx, updateMarketPricesMsg.MarketPriceUpdates)
```

Comment:
Why are we updating market prices in `ExtendVote`?

Reply:
Because VEs are generated before deliver tx runs. 

When we tested the protocol changes initially we ran into the following scenario:
1. ExtendVote in block `N-1` runs and suggests a price change to price `New`.
2. Prepare/Process Proposal runs in block `N` and the price change is accepted into the block.
3. ExtendVote runs in block `N` and suggests to change the price to `New` since DeliverTx hasn't actually changed the x/prices state.
4. Block `N+1` runs and PrepareProposal creates an `UpdateMarketPrices` Tx which is rejected in ProcessProposal because of the `min_price_change` constraint being violated.

The chain in this case would perpetually propose an UpdateMarketPrices tx which would be rejected and it would never make progress.

The solution to this was to execute the changes to the PricesKeeper in `ExtendVote` before calling `GetValidMarketPriceUpdates`. This way all suggested price changes are based on the state the prices will be in after the current block is applied.

Reply:
IIuc, Since ExtendVote for block `N` occurs before block `N` is committed, the price updates proposed by block `N` may not be valid? 

I think I'm a little confused too, because I thought if in Block `N+1` received a consensus through vote extension which `New + Delta` where `Delta` is less than `min_price_change`, PrepareProposal would just ignore that price and would not create an `UpdateMarketPrices` Tx  with a price update

Reply:
> IIuc, Since ExtendVote for block N occurs before block N is committed, the price updates proposed by block N may not be valid?

Correct. The changes are valid as long as we first apply the changes in block N. 

> I think I'm a little confused too, because I thought if in Block N+1 received a consensus through vote extension which New + Delta where Delta is less than min_price_change, PrepareProposal would just ignore that price and would not create an UpdateMarketPrices Tx with a price update

No, the [Default](https://github.com/skip-mev/v4-chain/blob/9cf8a678503243efa2f98af0fd49291121c44b7f/protocol/app/prepare/prices/default_price_update_generator.go#L31) prepare handler which uses the existing logic calls the Prices keeper to generate the tx which means prices with less than min_price_change movement are ignored.


In the [Slinky](https://github.com/skip-mev/v4-chain/blob/9cf8a678503243efa2f98af0fd49291121c44b7f/protocol/app/prepare/prices/slinky_price_update_generator.go#L66) prepare handler we just aggregate over the data in the VEs so the responsibility for making sure that data is valid is in the ExtendVoteHandler.

Reply:
We have prices transaction as the last one in a block as things such as order matches are created based on old prices. If we update prices in ExtendVote, that will potentially make other transactions invalid